### ISSUE TITLE
Rake as a library: REDUCE_COMPAT is unsuitable

### ISSUE THREAD

**Issue by kschiess** [jimweirich/rake#81]
To be able to specify which version of rake I use, rake figures as 'gem "rake"' in my Gemfile. That means that rake is part of my applications environment (which is useful when using app-related classes in the Rakefile as well). 

With rake, I inherit 'dsl_definition.rb' and the line 

```
include Rake::DeprecatedObjectDSL unless defined? Rake::REDUCE_COMPAT
```

To avoid getting rake methods in my objects I would have to 

```
module Rake
  REDUCE_COMPAT = true
end
```

before any kind of bundler setup. This can be done, but will be real hard in the case of 'bundler exec X'. The '--reduce-compat' switch doesn't help at all.

The same kind of objection applies to any kind of usage of rake as library. The bad thing here is really the namespace pollution that rake does, be it conditionally or unconditionally. #63 doesn't solve the problem at all. 

The default should be not to pollute global namespace, with the exception being the pollution, not the other way around. I realize this may break stuff, but we can safely assume that people can control their gem versions by now...

**Comment by floere** [comment]
+1. The option should be to include the deprecated DSL.

**Comment by the8472** [comment]
Workaround:

``` ruby
Rake::DeprecatedObjectDSL.private_instance_methods.each do |m|
  Rake::DeprecatedObjectDSL.send("undef_method",m)
end
```

**Comment by damphyr** [comment]
+1. I know it will break all kinds of things (including most of my own rakefiles) but for standalone rakefiles it's an easy fix and  it would make embedding rake in an app that much cleaner.

**Comment by kschiess** [comment]
@the8472 I have something like this in my projects. Hideous.

**Comment by leehambley** [comment]
I'm giving this a +1 as I'm just about to release Capistrano 2.10.0, and had to include the above hack. We create methods such as `symlink` and  `disable`, but we do it at runtime, isolated to a Capfile process space. We can't rely on the load order (quite the contrary, we can almost always guarantee that Rails/Bundler has already loaded Rake) being such that we can use `REDUCE_COMPAT`.

**Comment by kschiess** [comment]
For the record, this is what I'm using currently. Just spent another two hours on this: 

```
 class Module
  def with_rake_pollution_prevention_append_features(klass)
    unless caller.first.match(/rake/) && self.name == "Rake::DeprecatedObjectDSL"
      without_rake_pollution_prevention_append_features(klass)
    end
  end
  alias without_rake_pollution_prevention_append_features append_features
  alias append_features with_rake_pollution_prevention_append_features
end
```

**Comment by jimweirich** [comment]
We will be removing all the deprecated namespaces, method names, etc. post 0.9.x.  My goal is to release a 0.9.3 version with as many bug fixes as I can manage, then cleanup  the namespaces and release a version 10.0.x.

This issue is marked to be addressed post 0.9.x.

**Comment by jimweirich** [comment]
Fixed in 10.0.
